Add Compression Streams Brotli

stwrt opened the pull request:
#### 0d3451c531c810977123d2b9503ed968381d2004
<pre>
Add Compression Streams Brotli
<a href="https://bugs.webkit.org/show_bug.cgi?id=280445">https://bugs.webkit.org/show_bug.cgi?id=280445</a>
<a href="https://rdar.apple.com/99159497">rdar://99159497</a>

Reviewed by Alex Christensen and Ryosuke Niwa.

Add Brotli support to Compression Streams by using Apple&apos;s Compression Framework.

Co-authored-by: Alex Christensen &lt;achristensen@apple.com&gt;

Initial implementation and tests written by Brandon Stewart.
Alex implemented a substantial refactor to improve upon the implementation.

* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.js:
(const.chunk.of.badChunks.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.js:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.js:
(const.chunk.of.badChunks.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.worker-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.serviceworker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.serviceworker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.sharedworker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.sharedworker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.worker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.worker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.serviceworker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.serviceworker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.sharedworker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.sharedworker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.worker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.worker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.serviceworker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.serviceworker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.sharedworker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.sharedworker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.worker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.worker-expected.txt.
* Source/WebCore/Modules/compression/CompressionStream.cpp: Added.
(WebCore::CompressionStream::CompressionStream):
(WebCore::CompressionStream::~CompressionStream):
(WebCore::CompressionStream::initializeIfNecessary):
* Source/WebCore/Modules/compression/CompressionStream.h: Added.
(WebCore::CompressionStream::get):
* Source/WebCore/Modules/compression/CompressionStream.js:
(initializeCompressionStream):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
(WebCore::CompressionStreamEncoder::didDeflateFinish const):
(WebCore::CompressionStreamEncoder::compress):
(WebCore::compressionAlgorithm):
(WebCore::CompressionStreamEncoder::compressZlib):
(WebCore::CompressionStreamEncoder::initialize): Deleted.
* Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
(WebCore::CompressionStreamEncoder::create):
(WebCore::CompressionStreamEncoder::CompressionStreamEncoder):
(WebCore::CompressionStreamEncoder::~CompressionStreamEncoder): Deleted.
* Source/WebCore/Modules/compression/DecompressionStream.js:
(initializeDecompressionStream):
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
(WebCore::DecompressionStreamDecoder::decompress):
(WebCore::DecompressionStreamDecoder::didInflateFinish const):
(WebCore::DecompressionStreamDecoder::didInflateContainExtraBytes const):
(WebCore::decompressionAlgorithm):
(WebCore::DecompressionStreamDecoder::decompressZlib):
(WebCore::DecompressionStreamDecoder::initialize): Deleted.
(WebCore::DecompressionStreamDecoder::initializeAppleCompressionFramework): Deleted.
(WebCore::DecompressionStreamDecoder::decompressAppleCompressionFramework): Deleted.
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
(WebCore::DecompressionStreamDecoder::create):
(WebCore::DecompressionStreamDecoder::DecompressionStreamDecoder):
(WebCore::DecompressionStreamDecoder::~DecompressionStreamDecoder): Deleted.
* Source/WebCore/Modules/compression/Formats.h:
(): Deleted.
* Source/WebCore/Modules/compression/ZStream.cpp: Added.
(WebCore::ZStream::initializeIfNecessary):
(WebCore::ZStream::ZStream):
(WebCore::ZStream::~ZStream):
* Source/WebCore/Modules/compression/ZStream.h: Added.
(WebCore::ZStream::get):
(WebCore::ZStream::get const):
* Source/WebCore/Modules/compression/cocoa/CompressionStreamEncoderCocoa.mm: Added.
(WebCore::CompressionStreamEncoder::compressAppleCompressionFramework):
* Source/WebCore/Modules/compression/cocoa/DecompressionStreamDecoderCocoa.mm: Added.
(WebCore::DecompressionStreamDecoder::decompressAppleCompressionFramework):
* Source/WebCore/Modules/notifications/NotificationPayloadCocoa.mm:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/285391@main">https://commits.webkit.org/285391@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1535f2672269646376efbaf0586dabf5964c95b5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72443 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51864 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25231 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76624 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23652 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74558 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59668 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23474 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57040 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15560 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75510 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46957 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62390 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37476 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43608 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19856 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22002 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65490 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20214 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78292 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16684 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19346 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65507 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16732 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62402 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64774 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16010 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13037 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6680 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47662 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2447 "Found 1 new failure in Modules/webaudio/AudioNodeOutput.h and found 1 fixed file: Modules/screen-wake-lock/WakeLockManager.h") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48731 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50020 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48474 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
```diff
@@ -218,6 +229,7 @@ ExceptionOr<Ref<JSC::ArrayBuffer>> DecompressionStreamDecoder::decompressAppleCo
         m_stream.dst_size = output.size();
 
         result = compression_stream_process(&m_stream, m_didFinish ? COMPRESSION_STREAM_FINALIZE : 0);
+        WTF_ALWAYS_LOG(result);
```
Remove

achristensen07 commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
```diff
@@ -47,14 +50,21 @@ class CompressionStreamEncoder : public RefCounted<CompressionStreamEncoder> {
+        if (m_initialized) {
+            if (m_usingAppleCompressionFramework) {
+#if PLATFORM(COCOA)
+                compression_stream_destroy(&m_stream);
```
It would be nice if we had a smart-pointer-like thing that would take care of calling compression_stream_destroy at the proper time.

achristensen07 commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -109,6 +109,18 @@ static bool didDeflateFail(int result)
+    if (m_format == Formats::CompressionFormat::Brotli)
+        return compressAppleCompressionFramework(input);
+    else
+        return compressZlib(input);
```
Can we use the compression framework for zlib too?  That would make our abstractions better organized

stwrt replied:
Apple's Compression Framework did not fully support streaming of either `deflate`, `deflate-raw`, or `gzip`.

My memory is a bit fuzzy at the moment on which algorithm specifically was missing.

achristensen07 commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -164,4 +176,82 @@ ExceptionOr<Ref<JSC::ArrayBuffer>> CompressionStreamEncoder::compress(std::span<
     return compressedData.releaseNonNull();
 }
+
+
```
nit: extra space

achristensen07 commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -164,4 +176,82 @@ ExceptionOr<Ref<JSC::ArrayBuffer>> CompressionStreamEncoder::compress(std::span<
+
+
+#if PLATFORM(COCOA)
+ExceptionOr<bool> CompressionStreamEncoder::initializeAppleCompressionFramework() 
```
This kind of thing should probably go in a platform file like CompressionStreamEncoderCocoa.cpp

achristensen07 commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -164,4 +176,82 @@ ExceptionOr<Ref<JSC::ArrayBuffer>> CompressionStreamEncoder::compress(std::span<
+            output.shrink(allocateSize - m_stream.dst_size);
+        } else {
+            if (allocateSize < maxAllocationSize)
+                allocateSize *= 2;
```
This value is never used.

achristensen07 commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -164,4 +176,82 @@ ExceptionOr<Ref<JSC::ArrayBuffer>> CompressionStreamEncoder::compress(std::span<
+
+        if ((result == COMPRESSION_STATUS_END && m_stream.src_size)
+            || (m_didFinish && m_stream.src_size))
+            return Exception { ExceptionCode::TypeError, "Extra bytes past the end."_s };
```
I would expect this to result in retrying with a larger buffer then an OutOfMemoryError if it can't be allocated.

stwrt replied:
I think we may be able to remove this check from the compression section as this was designed to handle extraneous bytes added to the end of a compressed message. This check I believe is mainly relevant to decompression.

achristensen07 commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -164,4 +176,82 @@ ExceptionOr<Ref<JSC::ArrayBuffer>> CompressionStreamEncoder::compress(std::span<
+    while (shouldDecompress) {
+        Vector<uint8_t> output;
+        if (!output.tryReserveInitialCapacity(allocateSize)) {
+            allocateSize /= 4;
```
4 seems like a very large factor to shrink by if allocation fails.  That seems like it'll make it very likely to be too small.

achristensen07 commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
```diff
@@ -73,6 +83,14 @@ class CompressionStreamEncoder : public RefCounted<CompressionStreamEncoder> {
     bool m_didFinish { false };
     z_stream m_zstream;
 
+    bool m_usingAppleCompressionFramework { false };
```
This seems like it could be better organized by having something like a std::variant instead.

achristensen07 commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
```diff
@@ -73,6 +73,10 @@ class DecompressionStreamDecoder : public RefCounted<DecompressionStreamDecoder>
         std::memset(&m_zstream, 0, sizeof(m_zstream));
+
+#if PLATFORM(COCOA)
+        std::memset(&m_stream, 0, sizeof(m_stream));
```
It seems strange that the decoder zero initializes its members but the encoder doesn't.

achristensen07 commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
```diff
@@ -26,9 +26,11 @@
 #include "DecompressionStreamDecoder.h"
 
 #include "BufferSource.h"
+#include "wtf/Assertions.h"
```
Seems unnecessary.

achristensen07 commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
```diff
@@ -26,9 +26,11 @@
 #include <JavaScriptCore/GenericTypedArrayViewInlines.h>
 #include <JavaScriptCore/JSCInlines.h>
 #include <JavaScriptCore/JSGenericTypedArrayViewInlines.h>
+#include <compression.h>
```
Only include this from platform-specific places.

achristensen07 commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
```diff
@@ -27,6 +27,9 @@
 #include "BufferSource.h"
 #include "ExceptionOr.h"
 #include "Formats.h"
+#if PLATFORM(COCOA)
```
Yeah, having a separate implementation of the same thing lends itself to having a platform specific file.

stwrt replied:
I agree.

Currently prototyping a new version that adds new files for the zlib and compression streams backends.

achristensen07 commented on LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.js:
```diff
@@ -43,6 +43,7 @@ const chunkLists = [
 const expectedValue = new TextEncoder().encode('HelloHello');
 
 for (const chunkList of chunkLists) {
+
```
Whitespace

achristensen07 commented on Source/WebCore/Modules/compression/DecompressionStream.js:
```diff
@@ -27,12 +27,12 @@ function initializeDecompressionStream(format)
         @throwTypeError(errorMessage);
 
-    const algorithms = ['gzip', 'deflate', 'deflate-raw'];
+    const algorithms = ['brotli', 'gzip', 'deflate', 'deflate-raw'];
```
We probably only want to include brotli in the list of algorithms on platforms that support it.

stwrt replied:
A couple thoughts here.

1. Do we have a way to perform platform checks in JS (similar to `#if PLATFORM`)? 
2. We could move algorithms support checking to the C++ code where we can utilize the platform macros.

achristensen07 replied:
2

achristensen07 replied:
I took care of this another way, by returning an exception on non-Cocoa platforms when trying to create a brotli compressor or decompressor.

rreno commented on Source/WebCore/Modules/compression/Formats.h:
```diff
@@ -30,7 +30,7 @@ namespace WebCore {
 
 public:
-    enum CompressionFormat : uint8_t { Gzip, Zlib, Deflate };
+    enum CompressionFormat : uint8_t { Brotli, Gzip, Zlib, Deflate };
```
Any reason not to upgrade this to an `enum class`?

rreno commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -164,4 +176,82 @@ ExceptionOr<Ref<JSC::ArrayBuffer>> CompressionStreamEncoder::compress(std::span<
+#if PLATFORM(COCOA)
+ExceptionOr<bool> CompressionStreamEncoder::initializeAppleCompressionFramework() 
+{
+    m_initialized = true;
```
I think its possible to both fail to initialize the framework and still try to use it later by putting this here. I think this `m_initialized` variable shouldn't be set until after we've checked the return value from `compression_stream_init`.

rreno commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -164,4 +176,82 @@ ExceptionOr<Ref<JSC::ArrayBuffer>> CompressionStreamEncoder::compress(std::span<
+    if (result != COMPRESSION_STATUS_OK)
+        return Exception { ExceptionCode::TypeError, "Initialization Failed."_s };
+
+    return true;
```
We never return false so we don't really need 3 bits of data here. Why not just `ExceptionOr<void>`?

rniwa commented on Source/WebCore/Modules/compression/CompressionStream.h:
```diff
@@ -0,0 +1,57 @@
+    ~CompressionStream();
+
+#if PLATFORM(COCOA)
+    compression_stream& get()
```
get sounds super generic. How about getPlatformStream or something?

rniwa commented on Source/WebCore/Modules/compression/ZStream.h:
```diff
@@ -0,0 +1,48 @@
+
+    z_stream& get() { return m_stream; }
+    const z_stream& get() const { return m_stream; }
+private:
```
Blank line before `private:`?

rniwa commented on Source/WebCore/Modules/compression/CompressionStream.cpp:
```diff
@@ -0,0 +1,60 @@
+        return true;
+#if PLATFORM(COCOA)
+    auto result = compression_stream_init(&m_stream, operation == Operation::Compression ? COMPRESSION_STREAM_ENCODE : COMPRESSION_STREAM_DECODE, COMPRESSION_BROTLI);
+    if (result != COMPRESSION_STATUS_OK)
```
Hm... it's a bit odd that generic sounding CompressionStream only supports brotli.
Maybe it could be a trivial argument to this function? (i.e. enum class with just one value for now).

rniwa commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
```diff
@@ -25,42 +25,44 @@
 public:
-    static Ref<CompressionStreamEncoder> create(unsigned char format)
-    { 
+    static ExceptionOr<Ref<CompressionStreamEncoder>> create(unsigned char formatChar)
```
Why does this function take unsigned char, and not Formats::CompressionFormat?

achristensen07 replied:
That's a result of weird stream JS stuff.  I think we should address that in a separate PR.

rniwa replied:
oh, I see.

rniwa commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
```diff
@@ -69,10 +71,9 @@ class CompressionStreamEncoder : public RefCounted<CompressionStreamEncoder> {
-
-    Formats::CompressionFormat m_format;
+    ZStream m_zstream;
+    CompressionStream m_compressionStream;
```
Conceptually, this are mutually exclusive, right?
Maybe we can use `variant`? It's not a big deal to land this as is though.

stwrt replied:
Added `TODO` for work in a follow up patch.

stwrt committed:
Add Compression Streams Brotli
https://bugs.webkit.org/show_bug.cgi?id=280445
rdar://99159497

Reviewed by Alex Christensen and Ryosuke Niwa.

Add Brotli support to Compression Streams by using Apple's Compression Framework.

Co-authored-by: Alex Christensen <achristensen@apple.com>

Initial implementation and tests written by Brandon Stewart.
Alex implemented a substantial refactor to improve upon the implementation.

* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.js:
(const.chunk.of.badChunks.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.js:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.js:
(const.chunk.of.badChunks.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.worker-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.serviceworker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.serviceworker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.sharedworker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.sharedworker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.worker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.worker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.serviceworker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.serviceworker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.sharedworker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.sharedworker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.worker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.worker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.serviceworker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.serviceworker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.sharedworker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.sharedworker-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.worker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.worker-expected.txt.
* Source/WebCore/Modules/compression/CompressionStream.cpp: Added.
(WebCore::CompressionStream::CompressionStream):
(WebCore::CompressionStream::~CompressionStream):
(WebCore::CompressionStream::initializeIfNecessary):
* Source/WebCore/Modules/compression/CompressionStream.h: Added.
(WebCore::CompressionStream::get):
* Source/WebCore/Modules/compression/CompressionStream.js:
(initializeCompressionStream):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
(WebCore::CompressionStreamEncoder::didDeflateFinish const):
(WebCore::CompressionStreamEncoder::compress):
(WebCore::compressionAlgorithm):
(WebCore::CompressionStreamEncoder::compressZlib):
(WebCore::CompressionStreamEncoder::initialize): Deleted.
* Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
(WebCore::CompressionStreamEncoder::create):
(WebCore::CompressionStreamEncoder::CompressionStreamEncoder):
(WebCore::CompressionStreamEncoder::~CompressionStreamEncoder): Deleted.
* Source/WebCore/Modules/compression/DecompressionStream.js:
(initializeDecompressionStream):
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
(WebCore::DecompressionStreamDecoder::decompress):
(WebCore::DecompressionStreamDecoder::didInflateFinish const):
(WebCore::DecompressionStreamDecoder::didInflateContainExtraBytes const):
(WebCore::decompressionAlgorithm):
(WebCore::DecompressionStreamDecoder::decompressZlib):
(WebCore::DecompressionStreamDecoder::initialize): Deleted.
(WebCore::DecompressionStreamDecoder::initializeAppleCompressionFramework): Deleted.
(WebCore::DecompressionStreamDecoder::decompressAppleCompressionFramework): Deleted.
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
(WebCore::DecompressionStreamDecoder::create):
(WebCore::DecompressionStreamDecoder::DecompressionStreamDecoder):
(WebCore::DecompressionStreamDecoder::~DecompressionStreamDecoder): Deleted.
* Source/WebCore/Modules/compression/Formats.h:
(): Deleted.
* Source/WebCore/Modules/compression/ZStream.cpp: Added.
(WebCore::ZStream::initializeIfNecessary):
(WebCore::ZStream::ZStream):
(WebCore::ZStream::~ZStream):
* Source/WebCore/Modules/compression/ZStream.h: Added.
(WebCore::ZStream::get):
(WebCore::ZStream::get const):
* Source/WebCore/Modules/compression/cocoa/CompressionStreamEncoderCocoa.mm: Added.
(WebCore::CompressionStreamEncoder::compressAppleCompressionFramework):
* Source/WebCore/Modules/compression/cocoa/DecompressionStreamDecoderCocoa.mm: Added.
(WebCore::DecompressionStreamDecoder::decompressAppleCompressionFramework):
* Source/WebCore/Modules/notifications/NotificationPayloadCocoa.mm:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/285391@main
